### PR TITLE
SSR article

### PR DESCRIPTION
**Pull request by matthewp**
[Direct link](https://deploy-preview-67--astro-build.netlify.app/blog/experimental-server-side-rendering/)

### CODE REVIEW

File: src/pages/blog/experimental-server-side-rendering.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+No matter your approach, going client-side for authentication is always going to mean:
+
+- Things won’t run if JavaScript is disabled.
+- Your server can only respond with `200 OK` response codes.
+- It’s just plain slow. The user has to wait for the full HTML response (including JS, CSS, etc) even if they then immediately redirect. This creates multiple page requests before the user lands on the correct page.
+
+## Astro SSR: Designed to be Simple
+
+SSR might sound like a solved problem. Next.js, Nuxt, Gatsby, SvelteKit, and the whole gang of modern JavaScript meta-frameworks all have some idea of SSR already in place. SSR as a term is really just a rebranding of how Rails, PHP and other server frameworks already work today. If that’s all true, then what’s the big deal?
```
Comment (bholmesdev) [inline src/pages/blog/experimental-server-side-rendering.md:68]:
```suggestion
Next.js, Nuxt, Gatsby, SvelteKit, and the whole gang of modern JavaScript meta-frameworks have some idea of SSR already. So what makes Astro SSR so special?
```

Copied from the Notion feedback! Curious if we can shorten this lead-in to be more Astro-centric.

File: src/pages/blog/experimental-server-side-rendering.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+E-commerce is especially hurt by all three of the above problems. In an industry where every millisecond of load performance counts, how can you get the awesome performance of static HTML without the limitations?
+
+From Day 1, we knew that Astro’s approach to island architecture was uniquely positioned to solve this problem. Islands allow pages to load quickly and become interactive lazily; the user sees and is able to interact with the most important parts of the page, faster.
+
+# Enter: Server-Side Rendering (SSR)
+
+Server-side rendering is a well-understood solution to scaling large websites, going back to the earliest days of the internet. SSR is no silver bullet, but, when done right, SSR is an invaluable tool for certain use-cases.
+
+Take user authentication, which is something not easily achieved on a static site. You might do a couple of different ways:
```
Comment (bholmesdev) [inline src/pages/blog/experimental-server-side-rendering.md:37]:
```suggestion
Take user authentication for instance. With static site generation, you have a couple different approaches:
```

Nit: was unclear whether these approaches were the old "static site" way or the new SSR way.

File: src/pages/blog/experimental-server-side-rendering.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+```html
+<script>
+  if(!parseCookie(document.cookie).auth) {
+    window.location = '/login';
+  }
+</script>
+```
+
+1. Call an API that will check if a user is logged in, like `/api/auth`. Going this route is going to mean that you need to show loading spinners in each of your islands while you wait on the response.
```
Comment (bholmesdev) [inline src/pages/blog/experimental-server-side-rendering.md:49]:
```suggestion
2. Call an API that will check if a user is logged in, like `/api/auth`. Going this route is going to mean that you need to show loading spinners in each of your islands while you wait on the response.
```

File: src/pages/blog/experimental-server-side-rendering.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+From Day 1, we knew that Astro’s approach to island architecture was uniquely positioned to solve this problem. Islands allow pages to load quickly and become interactive lazily; the user sees and is able to interact with the most important parts of the page, faster.
+
+# Enter: Server-Side Rendering (SSR)
+
+Server-side rendering is a well-understood solution to scaling large websites, going back to the earliest days of the internet. SSR is no silver bullet, but, when done right, SSR is an invaluable tool for certain use-cases.
+
+Take user authentication, which is something not easily achieved on a static site. You might do a couple of different ways:
+
+1. Check for a cookie or JWT in localStorage, then redirect to a login page if the user doesn’t have one. If using a cookie this prevents using `HttpOnly`, opening you to malicious JavaScript attacks. Also not shown here is, what do you do if the token is not valid? You still need to check and respond in that case.
```
Comment (bholmesdev) [inline src/pages/blog/experimental-server-side-rendering.md:39]:
```suggestion
1. Check for a cookie or JWT in localStorage, then redirect to a login page if the user doesn’t have one. If using a cookie this prevents using `HttpOnly`, opening you to malicious JavaScript attacks.
```

Nit: this detail seemed to follow the code snippet. Maybe we can move this to the end?

File: src/pages/blog/experimental-server-side-rendering.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+```html
+<script>
+  if(!parseCookie(document.cookie).auth) {
+    window.location = '/login';
+  }
+</script>
+```
+
```
Comment (bholmesdev) [inline src/pages/blog/experimental-server-side-rendering.md:48]:
```suggestion

Additionally, what do you do if the token is not valid? You still need to check and respond in that case.

```

File: src/pages/blog/experimental-server-side-rendering.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+// 3. If no user was found, return a redirect. This instantly
+//    completes the response with the correct status code & headers.
+if(!user) {
+  return Astro.redirect('/login');
+}
+
+// 4. If the user is logged in, you can now use the `user` object
+//    right in your page template to show an avatar, name, etc.
+---
+<html> <!-- contents here --> </html>
```
Comment (bholmesdev) [inline src/pages/blog/experimental-server-side-rendering.md:97]:
```suggestion
<html>
...
<h1>Hello {user.name} 👋</h1>
<img src={user.avatar} alt={user.name}>
...
</html>
```

Nit: might be useful to show how simple Astro templating is. Some people with a nunjucks or jade background would appreciate it!

File: src/pages/blog/experimental-server-side-rendering.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+Modern web hosts set a high bar for developer experience. Developers expect frameworks to integrate with their favored providers, without too much configuration or setup.
+
+When we built Astro SSR, we evaluated the different ways that other frameworks tackle this problem. In the end, we decided to follow in the footsteps of [SvelteKit](https://kit.svelte.dev/) and [Remix](https://remix.run/) and adopt the idea of ***pluggable host*** ***adapters***. Adapters are simple, pluggable integrations that automatically configure your build for your favorite host.
+
+Not only is the adapter model easy for developers to set up, but it also lets us support as many different kinds of hosts as possible, including:
+
+- Run entirely on the edge with **Cloudflare** and **Deno Deploy**.
+- Run on a modern dev platform like **Netlify** and **Vercel**.
+- Run on a bare-metal serverless function on **AWS**, **Azure**, and **Google Cloud**.
+- Run on a JavaScript server runtime can deploy yourselfm like **Node.js** and **Deno.**
```
Comment (bholmesdev) [inline src/pages/blog/experimental-server-side-rendering.md:140]:
```suggestion
- Run on a JavaScript server runtime can deploy yourself like **Node.js** and **Deno.**
```

File: src/pages/blog/experimental-server-side-rendering.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+SSR support in Astro is still experimental. Over the next two months, we’re excited to work with all of the major hosting providers to launch more adapters and partnerships for every platform that our users care about. If you are a hosting provider interested in building your own Adapter for Astro, please reach out to us on Discord or via email: fred@astro.build.
```
Comment (bholmesdev) [inline src/pages/blog/experimental-server-side-rendering.md:142]:
```suggestion
SSR support in Astro is still experimental. Over the next two months, we’re excited to work with all of the major hosting providers to launch more adapters and partnerships for every platform that our users care about. If you are a hosting provider interested in building your own Adapter for Astro, please reach out to us on Discord or via email: partner@astro.build.
```

Leaving a note assuming this is ready!

**Review by bholmesdev — COMMENTED** [review]
Some comments on typos and clarity. Ready for everyone to be rendered speechless!

File: src/pages/blog/experimental-server-side-rendering.md (hunk 5 above)
Comment (bholmesdev) [reply src/pages/blog/experimental-server-side-rendering.md:48]: Moving line per previous comment ☝️

File: src/pages/blog/experimental-server-side-rendering.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
+---
+layout: '../../layouts/Post.astro'
+title: 'Server-side Rendering with Astro'
```
Comment (natemoo-re) [inline src/pages/blog/experimental-server-side-rendering.md:3]:
```suggestion
title: 'Server-Side Rendering with Astro'
```

**Review by natemoo-re — APPROVED** [review]
Just a small nit about title case being "Server-Side Rendering"

**Pull request merged**